README
NAME
    Dist::Zilla::Util::EmulatePhase - Nasty tools for probing Dist::Zilla's
    internal state.

VERSION
    version 1.001002

QUICK REFERENCE
      ::deduplicate(list=[])          # ArrayRef
      ::expand_modname(options=[])
        0           =>  $shortname
      ::get_plugins(options={})       # ArrayRef[Object]
        ^zilla      =>
        ?with       =>  [rolelist]
        ?skip_with  =>  [rolelist]
        ?isa        =>  [isalist]
        ?skip_isa   =>  [isalist]
      ::get_metadata(options={})      # HashRef
        ^zilla      =>
        ?with       =>  [rolelist]
      ::get_prereqs(options={})       # HashRef
        ^zilla      =>
        ?with       =>  [rolelist]

METHODS
  deduplicate
    Internal utility that de-duplicates references by ref-addr alone.

      my $array = [];
      is_deeply( [ deduplicate( $array, $array ) ],[ $array ] )

  expand_modname
    Internal utility to expand various shorthand notations to full ones.

      expand_modname('-MetaProvider') == 'Dist::Zilla::Role::MetaProvider';
      expand_modname('=MetaNoIndex')  == 'Dist::Zilla::Plugin::MetaNoIndex';

  get_plugins
    Probe Dist::Zilla's plugin registry and get items matching a
    specification

      my @plugins = get_plugins({
        zilla     => $self->zilla,
        with      => [qw( -MetaProvider -SomethingElse     )],
        skip_with => [qw( -SomethingBadThatIsAMetaProvider )],
        isa       => [qw( =SomePlugin   =SomeOtherPlugin   )],
        skip_isa  => [qw( =OurPlugin                       )],
      });

  get_metadata
    Emulates Dist::Zilla's internal metadata aggregation and does it all
    again.

    Minimum Usage:

      my $metadata = get_metadata({ zilla => $self->zilla });

    Extended usage:

      my $metadata = get_metadata({
        $zilla = $self->zilla,
         ... more params to get_plugins ...
         ... ie: ...
         with => [qw( -MetaProvider )],
         isa  => [qw( =MetaNoIndex )],
       });

  get_prereqs
    Emulates Dist::Zilla's internal prereqs aggregation and does it all
    again.

    Minimum Usage:

      my $prereqs = get_prereqs({ zilla => $self->zilla });

    Extended usage:

      my $metadata = get_prereqs({
        $zilla = $self->zilla,
         ... more params to get_plugins ...
         ... ie: ...
         with => [qw( -PrereqSource )],
         isa  => [qw( =AutoPrereqs )],
       });

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Dist-Zilla-Util-EmulatePhase

1.001003 2017-03-05T06:57:39Z e735bad

1.001002 2017-03-05T06:55:07Z 999a97a
 [Bugfix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Dist-Zilla-Util-EmulatePhase/issues/2

 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1 (recommends: +1)
 - test: +2 -1 (recommends: +1 ↑1 ↓1)

 [Tests]
 - Avoid use of dztest

1.001001 2014-08-29T09:51:56Z
 [Bugfix]
 - Avoid failures on old Path::Tiny stuff.

 [Dependencies::Stats]
 - Dependencies changed since 1.001000, see misc/*.deps* for details
 - test: ↑1

1.001000 2014-08-25T21:17:31Z
 [00 Minor]
 - Test refactoring.
 - Internals refactoring

 [Dependencies::Stats]
 - Dependencies changed since 1.000002, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - runtime: -1
 - test: +1 ↑1

 [Internals]
 - Code rewritten to not use Moose::Autobox while keeping semantics the same.

 [Tests]
 - Tests split
 - Tests refactored in terms of new dztest util.

1.000002 2014-08-15T22:05:57Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - augment tests.
 - update metadata

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - show 1x x_contributor

 [Tests]
 - update Test::ReportPrereqs

1.000001 2014-07-17T11:57:21Z
 [00 Trivial]
 - Misc packaging tidy ups.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - test: (recommends: ↑1)

1.000000 2014-04-18T10:46:37Z
 [00 Minor]
 - No signficant changes in the code, but potententially significant packaging changes downstream may care about.

 [01 Version Scheme Change]
 - Mantissa is moving from 8 to 6 to standardise.
 - This is the primary reason for 1.xxx as 0.xxxx is unambiguously smaller than 1.xxx

 [02 Toolkit Change]
 - ExtUtils::MakeMaker is now used instead of Module::Build
 - Tooling and Testing modules only now depend on the minimum required
 - and the "Latest" versions as per myself are only recommended upgrades.

 [Dependencies::Aggregate]
 - 5.008 : 184 → 201 : Details https://gist.github.com/kentfredric/11037169
 - 5.010 : 163 → 184 : Details https://gist.github.com/kentfredric/11036813
 - 5.019 : 159 → 162 : Details https://gist.github.com/kentfredric/11036128

 [Dependencies::Stats]
 - Dependencies changed since 0.01025803, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +41 ↑1 -1
 - runtime: +1 ↑1
 - test: +4 ↓1 -3 (recommends: +4)

0.01025803 2013-10-31T12:44:10Z
 [Bugfix]
 - Add MakeMaker::Awesome to the whitelist ( fixes RT#89828 )

 [Dependencies::Stats]
 - Dependencies changed since 0.01025802, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +8 -1 (recommends: -1, suggests: +1 -1)
 - test: +3 ↑1 -2

 [Documentation]
 - Specify encoding
 - add README.mkdn to dist
 - add MetaPOD Annotations
 - add MetaPOD layout

 [Meta]
 - bugs to github issues
 - travis testing now active

 [Packaging]
 - Use recent Module::Build for test_requires

 [Tests]
 - Regenerate 00-compile.t 
 - Regenerate 000-report-versions-tiny.t
 - Regenerate xt/release/cpan-changes.t
 - Regenerate xt/release/kwalitee.t

0.01025802 2012-07-28T20:17:35Z
 [Bugfix]
 - Handle some odd cases that turn up when you're working with scary deep nested code.
 - Handle having a faked zilla passed down to us instead of a real zilla ( ie: when another emulate::phase instance is
   triggering code triggering emulate phase )
 - Prereq collector now has a plugins() method that returns an empty array ref.
 - Can now actually bootstrap if deps are updated.

 [Dependencies::Stats]
 - Dependencies changed since 0.01025801, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: (recommends: ↑1, suggests: ↑1)
 - test: ↑1

 [Internals]
 - Code has been run through perltidy.

0.01025801 2012-01-04T02:01:25Z
 [Bugfixes]
 - rt#73546 Reported by MITHALDU/Christian Walde with initial patch by APOCAL. Thanks.

 [Dependencies::Stats]
 - Dependencies changed since 0.01025800, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +3
 - test: -1

 [Interals]
 - Declares AUTHORITY in all namespaces.
 - VERSION declaration now not in BEGIN

 [Packaging]
 - Migrate to Git Versions
 - Update copyright years/notice formatting.
 - Author/Release tests now in xt/
 - Git repository metadata now all https://

 [Tests]
 - Expect CPAN::Meta::Requirements instead of Version::Requirements
 - Updated: compile tests.

0.01025800 2011-07-13T02:16:46Z
 [Features]
 - The warning diagnostic call that triggers when plugins call '_share_dir_map' from register_plugins  has been
   special-cased with a white-list so that no errors are produced for the MakeMaker plugin and everything else will
   Carp fatally. Hopefully this doesn't prove a problem for anyone.

 [Packaging]
 - Updated copyright notice.
 - METAData updated for GitHub

0.01016513 2011-04-13T15:09:11Z
 - Maintenance only release.

 [Dependencies::Stats]
 - Dependencies changed since 0.01003312, see misc/*.deps* for details
 - develop: +1 -1 (recommends: +1 -1, suggests: +1 -1)

 [Packaging]
 - Move to @Author::KENTNL
 - Ship perltidyrc
 - Rework Changes for CPAN::Changes.

 [Tests]
 - Drop portability tests
 - Add CPAN::Changes tests

0.01003312 2010-12-05T14:48:13Z
 - Permit MakeMaker to call zilla->_share_dir_map on our shadow-zilla during pre-req collection. It should preferably
   be avoided, but this is needed for now to make it work. ( Reported by Apocalypse on IRC ).

0.01000515 2010-11-08T17:06:39Z
 - Depend on newer Test::Simple because subtest() can work wrong without it.
 - Depend on newer Util::Test::KENTNL


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4007"
  },
  "configure_requires" => {
    "Module::Build" => "0.4007"
  },
  "dist_abstract" => "Nasty tools for probing L<< C<Dist::Zilla>'s|Dist::Zilla >> internal state.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Dist-Zilla-Util-EmulatePhase",
  "dist_version" => "0.01025803",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Util::EmulatePhase",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Dist::Zilla::Prereqs" => 0,
    "Hash::Merge::Simple" => 0,
    "Moose" => 0,
    "Moose::Autobox" => 0,
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "Try::Tiny" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dist::Zilla::Util::Test::KENTNL" => "0.01000510",
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Fatal" => "0.003",
    "Test::More" => "0.99"
  }
);


my %fallback_build_requires = (
  "Dist::Zilla::Util::Test::KENTNL" => "0.01000510",
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.4007",
  "Test::Fatal" => "0.003",
  "Test::More" => "0.99"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use 5.008;    #utf8
use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    metadata KiokuDB KENTNL de deduplicate addr plugin prereq prereqs behaviour
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Nasty tools for probing Dist::Zilla's internal state.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Util-EmulatePhase",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Dist::Zilla::Util::EmulatePhase",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dist::Zilla::Prereqs" => 0,
    "Hash::Merge::Simple" => 0,
    "Moose" => 0,
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "Try::Tiny" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Path::Tiny" => 0,
    "Test::DZil" => 0,
    "Test::Fatal" => "0.003",
    "Test::More" => "0.96"
  },
  "VERSION" => "1.001003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Dist::Zilla::Prereqs" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Hash::Merge::Simple" => 0,
  "Moose" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Sub::Exporter" => 0,
  "Test::DZil" => 0,
  "Test::Fatal" => "0.003",
  "Test::More" => "0.96",
  "Try::Tiny" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
